cro-clib/src/wallet.rs

```diff
@@ -0,0 +1,241 @@
+    mnemonics: *mut u8,
+    mnemonics_length: u32,
+) -> CroResult {
+    let mnemonic = Mnemonic::new();
+    let phrase = mnemonic.unsecure_phrase();
+    assert!(phrase.as_bytes().len() < mnemonics_length as usize);
+    let wallet = CroHDWallet {
+        seed: HDSeed::from(&mnemonic),
+    };
+    let wallet_box = Box::new(wallet);
+    ptr::write(wallet_out, Box::into_raw(wallet_box));
+    ptr::copy_nonoverlapping(
+        phrase.as_bytes().as_ptr(),
+        mnemonics,
+        phrase.as_bytes().len(),
+    );
+    CroResult::success()
+}
+
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_restore_hdwallet(
+    mnemonics_string: *const c_char,
+    wallet_out: *mut CroHDWalletPtr,
+) -> CroResult {
+    let mnemonics = get_string(mnemonics_string);
+    let mnemonics_sec = SecUtf8::from(mnemonics);
+    let mnemonic = Mnemonic::from_secstr(&mnemonics_sec).unwrap();
+    let wallet = CroHDWallet {
+        seed: HDSeed::from(&mnemonic),
+    };
+    let wallet_box = Box::new(wallet);
+    ptr::write(wallet_out, Box::into_raw(wallet_box));
+    CroResult::success()
+}
+
+/// create staking address from bip44 hdwallet
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_create_staking_address(
+    wallet_ptr: CroHDWalletPtr,
+    network: CroNetwork,
+    address_out: *mut CroAddressPtr,
+    index: u32,
+) -> CroResult {
+    if wallet_ptr.is_null() {
+        return CroResult::fail();
+    }
+    let wallet = wallet_ptr.as_mut().expect("get wallet");
+    let (public, private) = wallet
+        .seed
+        .derive_key_pair(network.into(), CroAccount::Staking as u32, index)
+        .unwrap();
+    let address = StakedStateAddress::BasicRedeem(RedeemAddress::from(&public));
+
+    match address {
+        StakedStateAddress::BasicRedeem(redeem) => {
+            assert!(20 == redeem.0.len());
+            let raw = redeem.to_vec();
+
+            let ret = CroAddress {
+                privatekey: private,
+                publickey: public,
+                raw,
+                address: address.to_string(),
+            };
+            let address_box = Box::new(ret);
+            ptr::write(address_out, Box::into_raw(address_box));
+            CroResult::success()
+        }
+    }
+}
+
+/// print address information
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_print_address(address_ptr: CroAddressPtr) -> CroResult {
+    let address = address_ptr.as_mut().expect("get address");
+    println!("{}", address.address.to_string());
+    CroResult::success()
+}
+
+/// print address information
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_get_printed_address(
+    address_ptr: CroAddressPtr,
+    address_output: *mut u8,
+    address_output_length: u32,
+) -> CroResult {
+    let address = address_ptr.as_mut().expect("get address");
+    let address_string = address.address.to_string();
+    let mut src_string = address_string.as_bytes().to_vec();
+    src_string.push(0);
```

Comment:
use `CString`: https://doc.rust-lang.org/std/ffi/struct.CString.html

Reply:
https://github.com/crypto-com/chain/pull/895#discussion_r367745914

Reply:
CString is owned by Rust, give C can write it,
but above api is writing string inside rust, which is owned in C

Reply:
I mean, you can still do ` ptr::copy_nonoverlapping` -- just that `src` will be `CString` instead of `Vec<u8>`

Reply:
ok, got it